File: updraft_sdk/src/main/java/com/apsswithlove/updraft_sdk/api/ApiWrapper.java
package com.apsswithlove.updraft_sdk.api;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import com.apsswithlove.updraft_sdk.BuildConfig;
import com.apsswithlove.updraft_sdk.Settings;
import com.apsswithlove.updraft_sdk.api.request.CheckLastVersionRequest;
import com.apsswithlove.updraft_sdk.api.request.FeedbackMobileRequest;
import com.apsswithlove.updraft_sdk.api.request.GetLastVersionRequest;
import com.apsswithlove.updraft_sdk.api.response.CheckLastVersionResponse;
import com.apsswithlove.updraft_sdk.api.response.FeedbackMobileResponse;
import com.apsswithlove.updraft_sdk.api.response.GetLastVersionResponse;
import com.apsswithlove.updraft_sdk.feedback.form.FeedbackChoice;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Single;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.RequestBody;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by satori on 3/27/18.
 */

public class ApiWrapper {

    private UpdraftService mUpdraftService;
    private Context mContext;
    private Settings mSettings;

    public ApiWrapper(Context context, Settings settings) {
        mContext = context;
        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder();

        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        if (settings.getLogLevel() == Settings.LOG_LEVEL_DEBUG) {
            logging.setLevel(HttpLoggingInterceptor.Level.BODY);
        }
        if (settings.getLogLevel() == Settings.LOG_LEVEL_NONE || settings.getLogLevel() == Settings.LOG_LEVEL_ERROR) {
            logging.setLevel(HttpLoggingInterceptor.Level.NONE);
        }
        httpClientBuilder.addInterceptor(logging);

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(settings.getBaseUrl())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(httpClientBuilder.build())
                .build();
        mUpdraftService = retrofit.
                create(UpdraftService.class);
        mSettings = settings;
    }

    public Single<GetLastVersionResponse> getLastVersion() {
        return Single.defer(() -> {
            GetLastVersionRequest getLastVersionRequest = new GetLastVersionRequest();
            getLastVersionRequest.setSdkKey(mSettings.getSdkKey());
            getLastVersionRequest.setAppKey(mSettings.getAppKey());
            return mUpdraftService.getLastVersion(getLastVersionRequest);
        });
    }

    public Single<CheckLastVersionResponse> checkLastVersion() {
        return Single.defer(() -> {
            CheckLastVersionRequest checkLastVersionRequest = new CheckLastVersionRequest();
            checkLastVersionRequest.setAppKey(mSettings.getAppKey());
            checkLastVersionRequest.setSdkKey(mSettings.getSdkKey());
            int versionCode = -1;
            try {
                PackageInfo packageInfo = mContext.getPackageManager().getPackageInfo(mContext.getPackageName(), 0);
                versionCode = packageInfo.versionCode;
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
            if (versionCode == -1) {
                return Single.error(new IllegalStateException("Version code is invalid"));
            }
            checkLastVersionRequest.setVersion(Integer.toString(versionCode));
            return mUpdraftService.checkLastVersion(checkLastVersionRequest);
        });
    }

    public Observable<Double> sendMobileFeedback(FeedbackChoice feedbackChoice, String description, String email, String fileName) {

        return Observable.defer(() -> {
            FeedbackMobileRequest feedbackMobileRequest = new FeedbackMobileRequest();
            feedbackMobileRequest.setAppKey(mSettings.getAppKey());
            feedbackMobileRequest.setSdkKey(mSettings.getSdkKey());
            feedbackMobileRequest.setTag(feedbackChoice.getApiName());
            feedbackMobileRequest.setImage(fileName);
            feedbackMobileRequest.setDescription(description);
            feedbackMobileRequest.setEmail(email);
            PackageManager m = mContext.getPackageManager();
            String appVersion = "";
            try {
                appVersion = m.getPackageInfo(mContext.getPackageName(), 0).versionName;
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
            feedbackMobileRequest.setBuildVersion(appVersion);
            feedbackMobileRequest.setSystemVersion(Build.VERSION.RELEASE);
            feedbackMobileRequest.setDeviceName(Build.MODEL);
            String androidId = android.provider.Settings.Secure.getString(mContext.getContentResolver(),
                    android.provider.Settings.Secure.ANDROID_ID);
            feedbackMobileRequest.setDeviceUuid(androidId);
            return Observable.create(new ObservableOnSubscribe<Double>() {
                @Override
                public void subscribe(ObservableEmitter<Double> emitter) throws Exception {
                    try {
                        FeedbackMobileResponse response = mUpdraftService.feedbackMobile(mapMultipart(feedbackMobileRequest, (bytesWritten, contentLength) -> {
                            double progress = (1.0 * bytesWritten) / contentLength;
                            emitter.onNext(progress);
                        })).blockingGet();
                    } catch (Throwable t) {
                        if (mSettings.shouldShowErrors()) {
                            t.printStackTrace();
                        }
                        emitter.onError(t);
                    }
                    emitter.onComplete();
                }
            });
        });
    }

    private Map<String, RequestBody> mapMultipart(FeedbackMobileRequest request, CountingRequestBody.Listener listener) {
        Map<String, RequestBody> map = new HashMap<>();
        File imageFile = new File(mContext.getFilesDir(), request.getImage());
        RequestBody imageRequestBody = new CountingRequestBody(RequestBody.create(MediaType.parse("image/png"), imageFile), listener);
        MediaType plainText = MediaType.parse("text/plain");
        map.put("image\"; filename=\"" + imageFile.getName(), imageRequestBody);
        if (request.getAppKey() != null) {
            map.put("app_key", RequestBody.create(plainText, request.getAppKey()));
        }
        if (request.getSdkKey() != null) {
            map.put("sdk_key", RequestBody.create(plainText, request.getSdkKey()));
        }
        if (request.getTag() != null) {
            map.put("tag", RequestBody.create(plainText, request.getTag()));
        }
        if (request.getDescription() != null) {
            map.put("description", RequestBody.create(plainText, request.getDescription()));
        }
        if (request.getEmail() != null) {
            map.put("email", RequestBody.create(plainText, request.getEmail()));
        }
        if (request.getBuildVersion() != null) {
            map.put("build_version", RequestBody.create(plainText, request.getBuildVersion()));
        }
        if (request.getSystemVersion() != null) {
            map.put("system_version", RequestBody.create(plainText, request.getSystemVersion()));
        }
        if (request.getDeviceName() != null) {
            map.put("device_name", RequestBody.create(plainText, request.getDeviceName()));
        }
        if (request.getDeviceUuid() != null) {
            map.put("device_uuid", RequestBody.create(plainText, request.getDeviceUuid()));
        }
        if (request.getNavigationStack() != null) {
            map.put("navigation_stack", RequestBody.create(plainText, request.getNavigationStack()));
        }
        return map;
    }

}

File: updraft-sdk/src/main/java/com/appswithlove/updraft/api/ApiWrapper.kt
package com.appswithlove.updraft.api

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import com.appswithlove.updraft.Settings
import com.appswithlove.updraft.api.request.CheckLastVersionRequest
import com.appswithlove.updraft.api.request.FeedbackEnabledRequest
import com.appswithlove.updraft.api.request.FeedbackMobileRequest
import com.appswithlove.updraft.api.request.GetLastVersionRequest
import com.appswithlove.updraft.api.response.CheckLastVersionResponse
import com.appswithlove.updraft.api.response.GetLastVersionResponse
import com.appswithlove.updraft.feedback.form.FeedbackChoice
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import java.io.File

class ApiWrapper(
    private val context: Context,
    private val settings: Settings
) {
    private val updraftService: UpdraftService

    init {
        val httpClientBuilder = OkHttpClient.Builder()

        val logging = HttpLoggingInterceptor().apply {
            level = when (settings.logLevel) {
                Settings.LOG_LEVEL_DEBUG -> HttpLoggingInterceptor.Level.BODY
                Settings.LOG_LEVEL_NONE, Settings.LOG_LEVEL_ERROR -> HttpLoggingInterceptor.Level.NONE
                else -> HttpLoggingInterceptor.Level.NONE
            }
        }
        httpClientBuilder.addInterceptor(logging)

        val json = Json { ignoreUnknownKeys = true }
        val retrofit = Retrofit.Builder()
            .baseUrl(settings.baseUrl)
            .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
            .client(httpClientBuilder.build())
            .build()

        updraftService = retrofit.create(UpdraftService::class.java)
    }

    suspend fun getLastVersion(): GetLastVersionResponse {
        val request = GetLastVersionRequest(
            sdkKey = settings.sdkKey,
            appKey = settings.appKey,
        )
        return updraftService.getLastVersion(request)
    }

    suspend fun checkLastVersion(): CheckLastVersionResponse {
        val request = CheckLastVersionRequest(
            appKey = settings.appKey,
            sdkKey = settings.sdkKey,
        )

        val versionCode = try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                packageInfo.longVersionCode.toInt()
            } else {
                @Suppress("DEPRECATION")
                packageInfo.versionCode
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            -1
        }

        if (versionCode == -1) throw IllegalStateException("Version code is invalid")

        return updraftService.checkLastVersion(request.copy(version = versionCode.toString()))
    }

    suspend fun isFeedbackEnabled(): Boolean {
        val request = FeedbackEnabledRequest(
            appKey = settings.appKey,
            sdkKey = settings.sdkKey,
        )

        val response = updraftService.isFeedbackEnabled(request)
        if (response.errorCodes.isNotEmpty()) {
            val errorDescription = response.errorDescriptions.firstOrNull().orEmpty()
            throw ApiException(errorDescription)
        }
        return response.isFeedbackEnabled
    }

    @SuppressLint("HardwareIds")
    fun sendMobileFeedback(
        feedbackChoice: FeedbackChoice,
        description: String,
        email: String,
        fileName: String
    ): Flow<Double> = callbackFlow {
        val request = FeedbackMobileRequest(
            appKey = settings.appKey,
            sdkKey = settings.sdkKey,
            tag = feedbackChoice.apiName(),
            image = fileName,
            description = description,
            email = email,
            buildVersion = try {
                context.packageManager.getPackageInfo(context.packageName, 0).versionName
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
                ""
            },
            systemVersion = Build.VERSION.RELEASE,
            deviceName = Build.MODEL,
            deviceUuid = android.provider.Settings.Secure.getString(
                context.contentResolver,
                android.provider.Settings.Secure.ANDROID_ID
            ),
        )

        val multipartMap = mapMultipart(request) { bytesWritten, contentLength ->
            val progress = bytesWritten.toDouble() / contentLength.toDouble()
            trySend(progress).isSuccess
        }

        val uploadJob = launch {
            try {
                updraftService.feedbackMobile(multipartMap)
                close()
            } catch (t: Throwable) {
                if (settings.shouldShowErrors()) t.printStackTrace()
                close(t)
            }
        }

        awaitClose {
            uploadJob.cancel()
        }
    }

    private fun mapMultipart(
        request: FeedbackMobileRequest,
        listener: CountingRequestBody.Listener
    ): Map<String, RequestBody> {
        val map = mutableMapOf<String, RequestBody>()
        val imageFile = File(context.filesDir, request.image ?: "")
        val imageRequestBody = CountingRequestBody(
            imageFile.asRequestBody("image/png".toMediaTypeOrNull()),
            listener,
        )
        val plainText = "text/plain".toMediaTypeOrNull()

        map["image\"; filename=\"${imageFile.name}"] = imageRequestBody

        request.appKey?.let { map["app_key"] = it.toRequestBody(plainText) }
        request.sdkKey?.let { map["sdk_key"] = it.toRequestBody(plainText) }
        request.tag?.let { map["tag"] = it.toRequestBody(plainText) }
        request.description?.let { map["description"] = it.toRequestBody(plainText) }
        request.email?.let { map["email"] = it.toRequestBody(plainText) }
        request.buildVersion?.let { map["build_version"] = it.toRequestBody(plainText) }
        request.systemVersion?.let { map["system_version"] = it.toRequestBody(plainText) }
        request.deviceName?.let { map["device_name"] = it.toRequestBody(plainText) }
        request.deviceUuid?.let { map["device_uuid"] = it.toRequestBody(plainText) }
        request.navigationStack?.let { map["navigation_stack"] = it.toRequestBody(plainText) }

        return map
    }
}
